Set language for scripted field

Bargs opened the pull request:
Fixes https://github.com/elastic/kibana/issues/6529
Based on the work started in #7206

This PR allows Kibana scripted fields to be written in any language that has been enabled for inline scripting in ES. By default, that's `expression` and `painless` in 5.0 (and technically mustache, but I don't expose it as an option since it makes no sense in script_fields). Groovy and other languages can be enabled in the Elasticsearch config. However, I've spent the most time trying to optimize the experience for `painless` and `expression` since those are the two sandboxed languages we should be encouraging people to use.

If you'd like to enable groovy in dev mode, you can add the following code to your grunt esvm (`esvm.js`) dev config block:

```
script: {
  engine: {
    groovy: {
      inline: true
    }
  }
}
```

Here's a screenshot of how things should look when it's running:

![screen shot 2016-08-19 at 4 01 49 pm](https://cloud.githubusercontent.com/assets/6239176/17822836/4c21c17e-6627-11e6-89b8-50388d8defaf.png)

Todo:
- [x] Test scripted fields with various field types
- [x] Restrict type options to those that actually work with scripted fields
- [x] Formatter list not getting updated when type is changed
- [x] Unit tests
- [x] API tests
- [x] Update in page docs
- [x] Date type scripted fields getting added to `fielddata_fields` in Discover
- [x] Restrict type based on language selected? (e.g. only numbers when expression lang is used)
- [x] Update `(Default: Number)` field label when type updates
- [x] Make painless the default lang

Scripted Field Types (from all possible Kibana field types):
string (yes)
date (yes, but needs fix for issue above)
boolean (yes)
number (yes)
geo_point (no, geohash_grid doesn't support scripts)
geo_shape (no)
ip (doesn't seem to work at the moment https://github.com/elastic/elasticsearch/issues/20067)
attachment (doesn't seem useful)
murmur3 (only comes with a plugin, probably not worth supporting OOB)

thomasneirynck commented:
This is somewhat of a nice to have, and can be split up from this PR.

But it would be helpful if the scripted field was validated when editing/creating it. Right now, you have to use it in a viz before you can tell if something's wrong.


e.g., following compilation error could be found at design time:
- either by testing it on the first document (less hard)
- inspecting the index mapping (harder)
![image](https://cloud.githubusercontent.com/assets/1833023/17874182/e78a0410-6897-11e6-8069-5ca1fd33bfa2.png)

Is there API in ES to test/compile a script before running it?

thomasneirynck commented:
I  like that the forms update automatically depending on the selection. E.g. the `expression` language only gives you the `number` as a possible return type, while with `painless` you have multiple options. It's a good safety mechanism, but it is not very intuitive when changing something on the bottom, changes input on the top.

Since these dependencies cascade down, I would move up the script-language selection to the top of the form, since it governs so much of what comes below (`Language` > `Type` > `Format`).

(You'll want to check this off with a UX expert too though. Apart from moving `language` to the top, I would even move up the `script`-field to the 2nd row. It's the most important field after all and determines if all the rest under it is valid).

thomasneirynck commented:
I'm not a fan of the `popularity` field in "scripted fields".

It's not explained with a warning-blurbs, or a tooltip. It is not explained in any of the three links on the page (`script fields`, `scripts in aggregations`, `lucene expressions`).

In the Kibana documentation, we can learn it is used to sort the fields in the field-list. But then I'm not 100% it actually works (see #5029).

Since we have such a nice free text filter search bar, and I don't expect scripted fields to exceed more than 25 (or perhaps they do (?)), I would just remove the option from the form altogether.

thomasneirynck commented on src/ui/public/index_patterns/_get_computed_fields.js:
```diff
@@ -6,7 +6,7 @@ export default function () {
   let docvalueFields = [];
 
-  docvalueFields = _.pluck(self.fields.byType.date, 'name');
+  docvalueFields = _.pluck(_.reject(self.fields.byType.date, 'scripted'), 'name');
```
FWIW, plucks deprecated in `lodash` from v4 onwards, so if we ever upgrade, we'll have to remove usage. (use 'map' instead).

thomasneirynck commented on src/core_plugins/kibana/server/routes/api/scripts/register_languages.js:
```diff
@@ -0,0 +1,27 @@
+    handler: function (request, reply) {
+      const boundCallWithRequest = _.partial(server.plugins.elasticsearch.callWithRequest, request);
+
+      return boundCallWithRequest('cluster.getSettings', {
```
Just curious, but why not just `server.plugins.elasticsearch.callWithRequest(request, 'cluster.getSettings`, { ...`

I've seen this pattern elsewhere in Kibana as well, but not quite sure why we create a new function.

Bargs replied:
I think I did it here purely out of habit, I'll switch it to using callWithRequest directly to make it less confusing.

Usually it's for improving testability of libraries that need to make calls to ES. Instead of having the lib require a hapi request object, and then having to mock out that request object, you just have to worry about mocking this one function that gets passed in. In this case we're not using any external libs though, so it doesn't make much sense.

thomasneirynck commented on src/ui/public/field_editor/field_editor.js:
```diff
@@ -129,6 +154,14 @@ uiModules
+        return $http.get(chrome.addBasePath('/api/kibana/scripts/languages'))
+        .then((res) => res.data)
+        .catch(() => {
+          return $q.reject(new Error('Error getting available scripting languages from Elasticsearch'));
```
Not sure how consistently Kibana uses these, but `ui/public/errors` has a lot of typed errors. You could reuse `RequestFailure` here.

Bargs replied:
Hah, you're the first person to ever point those out to me. Good to know!

Bargs replied:
Actually, those errors seem to be tailored to handling responses from the Elasticsearch proxy in Kibana. Using it with the response from this Kibana API gives a nasty "unknown" or "undefined" error, so I'm going to stick with with this custom message that's more user friendly. However I did update it to use the notifier that's already configured for this directive instead of simply rejecting with a new error.

Bargs commented:
@thomasneirynck I think I've addressed all your comments. A few replies:

* Script validation - agree, nice to have, but makes sense to do it as a separate enhancement. I don't think there's any script validation API in ES atm
* Language input location - totally agree, I've moved it to the top, just under the name input. I like having script at the bottom though, since it can expand in size, and it's where users are used to seeing it.
* popularity - it affects sorting in the Discover sidebar, and as far as I can tell it works. Agreed that it could be explained better, but I think that belongs in another PR

Please take another look!

thomasneirynck commented:
@Bargs 

- script validation: I'll log an enhancement for that
- language input location:  ok
- popularity: OK. I'm like Donny here, no frame of reference, like a child who wanders in the middle of a movie. The popularity field does work as intended indeed. What is "broken" (#5029) is that sort order isn't respected in the field editor-list, and after editing, fields are pushed to the back of the list making them "disappear".

OK, I'll look it over once more, thanks!

Bargs commented:
Just pushed a couple functional tests, and with that I've finished all the "code", pending any bugs or updates based on review of course. Just gotta work on the in page docs now.

thomasneirynck commented:
jenkins, test this.

Bargs commented:
I think there are unrelated test failures caused by the ESVM refresh last night :(

thomasneirynck commented:
LGTM

tbragin commented:
From a functional perspective, I tried a number of Lucene expressions and Painless scripted fields and so far these worked well. Working with the [getting-started dataset](https://www.elastic.co/guide/en/kibana/5.0/getting-started.html) for testing purposes. Capturing them here in case other testers want ideas for things to try, or for potential examples to include in the docs.

**Lucene expression - Number math on a field**
`doc['bytes'].value / 1024`

**Lucene expression - Parse date into DOW, etc.. (into number)**
`doc['@timestamp'].date.dayOfWeek`

**Painless - Combine two string values**
`doc['machine.os.keyword'].value + ' ' + doc['extension.keyword'].value`

**Painless - Match a string using regex, and take action on a match**
`if (doc['referer.keyword'].value =~ /error/) { return "error"} else {return "no error"}`

**Painless - Match a string and return that match**
`def m = /^.*\.([a-z]+)$/.matcher(doc['host.keyword'].value); if ( m.matches() ) { return m.group(1) } else { return "no match" }`

**Painless - Parse date into DOW, etc.. (into string)**
`LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['@timestamp'].value), ZoneId.of('Z')).getDayOfWeek().getDisplayName(TextStyle.FULL, Locale.getDefault());`

<img width="1174" alt="screen shot 2016-08-23 at 9 54 26 am" src="https://cloud.githubusercontent.com/assets/933397/17901295/9f15ba6e-6917-11e6-86ec-1177938057f3.png">

tbragin commented:
From a usability perspective, it would be nice if the script field box automatically expanded to the size of the script, since painless scripts tend to be longer. If that's a reasonable idea, but should be handled as part of a separate PR, happy to file a ticket. 

Default view:
<img width="1169" alt="screen shot 2016-08-23 at 10 14 16 am" src="https://cloud.githubusercontent.com/assets/933397/17902053/89b39a8a-691a-11e6-8e16-dc2a1a434d8c.png">

Manually expanded:
<img width="1172" alt="screen shot 2016-08-23 at 10 14 25 am" src="https://cloud.githubusercontent.com/assets/933397/17902052/897cbe5c-691a-11e6-9e79-981fcf5a6cd9.png">

I also agree with @thomasneirynck and @ppf2 on the fact that having additional validation and preview of script effects on a sample document would help from a workflow development perspective, as a future enhancement:
https://github.com/elastic/kibana/issues/8057
https://github.com/elastic/kibana/issues/2781

tsullivan commented on src/ui/public/field_editor/field_editor.html:
```diff
@@ -94,7 +110,48 @@ <h4 class="hintbox-heading">
+        </h4>
+
+        <p>
+          By default, Kibana scripted fields use <a target="_window" ng-href="{{editor.docLinks.painless}}">Painless <i class="fa-link fa"></i></a>, a simple and secure scripting language designed specifically for use with Elasticsearch.  To access values in the document use the following format:
```
I'm getting a 404 errors for all the links in this section: Painless doc: https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-painless.html 

Native Java APIs: https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-painless.html#painless-api

Painless syntax: https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-painless-syntax.html

Lucene expressions: https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-expression.html

Bargs replied:
Sorry, I was rushing to get that last commit in, meant to add a note about that. These should be valid links once an actual 5.0 branch exists in ES. ^ That's still correct, isn't it @clintongormley?

clintongormley replied:
Yes, but if you merge this change before we have a 5.0 branch it will break the docs build.

clintongormley replied:
I'll disable the kibana link checking until the ES has a 5.0 branch.

clintongormley replied:
Done

Bargs replied:
Thank you @clintongormley!

nik9000 commented:
> ```def m = /^.*\.([a-z]+)$/.matcher(doc['host.keyword'].value); if ( m.matches() ) { return m.group(1) } else { return "no match" }```

I think we should push people to put new lines in their scripts with examples like:

```groovy
def m = /^.*\.([a-z]+)$/.matcher(doc['host.keyword'].value);
if (m.matches()) {
  return m.group(1);
} else {
  return "no match";
}

I like new lines....

tbragin commented:
Another thing that might be useful is ability to disable scripted fields -- say if one is slow or isn't working. Right now the only way to stop them from running is to delete them.

tbragin commented:
@Bargs I checked out the docs in the latest commit and I like what you did there from a wording perspective. 

However, I noticed the links redirect to different places -- the first two to "current" (which currently redirects to the 2.3 docs) and the last four to "5.0" (which currently results in a 404).

https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-script-fields.html#search-request-script-fields

https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#search-aggregations-bucket-terms-aggregation-script

https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-painless.html

https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-painless.html#painless-api

https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-painless-syntax.html

https://www.elastic.co/guide/en/elasticsearch/reference/5.0/modules-scripting-expression.html

Bargs commented:
Ah great catch @tbragin, I missed those first two because I didn't modify the warning message at all. I've updated those first two to use the same dynamic doc linking strategy as the others. They'll 404 at the moment because there's no ES 5.0 branch (see comments from Clint in the inline comments above). You can validate that the rest of the URL is correct by replacing `5.0` with `master` in the URL.

tsullivan commented on src/core_plugins/kibana/server/routes/api/scripts/register_languages.js:
```diff
@@ -0,0 +1,27 @@
+import _ from 'lodash';
+import handleESError from '../../../lib/handle_es_error';
+
+export function registerLanguages(server) {
```
not an issue, but I'm wondering why it's not an `export default` here

Bargs replied:
Not really sure, just following the pattern of the other route registration modules in this plugin.

tsullivan commented on test/unit/api/scripts/_languages.js:
```diff
@@ -0,0 +1,27 @@
+        return request.get('/kibana/scripts/languages')
+        .expect(200)
+        .then(function (response) {
+          expect(response.body).to.be.an('array');
```
I haven't looked much at this kind of test code before. Does it actually query the server connected to Elasticsearch and get an actual response, or are there mocks somewhere?

Bargs replied:
No mocks, these are full integration tests.

tsullivan commented:
code LGTM, just had a few questions

Bargs commented:
@tsullivan @thomasneirynck @tbragin I made one more update to show an error message and disable the scripted field creation form when inline scripting is disabled in ES:

![screen shot 2016-08-24 at 7 33 11 pm](https://cloud.githubusercontent.com/assets/6239176/17951778/d6bed56a-6a32-11e6-824d-a7f401d17ced.png)

Could you all take one more quick look?

There [appears to be a bug](https://github.com/elastic/kibana/issues/6529#issuecomment-242231219) in ES's reporting of enabled languages when you simply disable all scripting, so to test for now you'll need to disabled `painless` and `expression` individually

You can do that like this in your grunt esvm.js file:

```
    dev: {
      options: {
        directory: resolve(directory, 'dev'),
        config: {
          path: {
            data: dataDir
          },
          cluster: {
            name: 'esvm-dev'
          },
          script: {
            engine: {
              painless: {
                inline: false
              },
              expression: {
                inline: false
              }
            }
          }
        }
      }
    },
```

Bargs commented:
jenkins, test this

Bargs commented:
jenkins, test this

Bargs commented:
Tests finally passed, this is g2g once I get a final LGTM on that last change I made

thomasneirynck commented:
LGTM

For future consideration: I think that warning message would also be appropriate to put on the "Scripted Fields" tab of Index-pattern overview, since it is a global setting.

Bargs commented:
Thanks @thomasneirynck! I'm gonna squash and merge Monday morning.

epixa commented:
Is that unchecked todo item still pending?

Bargs commented:
I couldn't reproduce it, I think it may have been an unrelated issue that got fixed elsewhere. I mostly left it there to remind myself to keep an eye out, but I guess I can remove it at this point.
